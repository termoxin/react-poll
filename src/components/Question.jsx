import React, { Component } from "react";
import { Form, Card, Icon, Radio, Button } from "semantic-ui-react";
import styled from "styled-components";
import QuestionType from "./QuestionType";

const ButtonContent = styled(Card.Content)`
  display: flex;
  justify-content: flex-end;
`;

const StyledCard = styled(Card)`
  && {
    width: 768px;
    min-width: 320px;
  }
`;

const StyledForm = styled(Form)`
  display: flex;
  flex-flow: column nowrap;
  align-items: flex-start;
`;

class Question extends Component {
  constructor(props) {
    super(props);

    this.state = {
      value: "",
      status: null,
      disabled: false
    };
  }

  handleChange = (e, { value, checked }) => {
    const { type } = this.props;
    const { status } = this.state;

    if (type === "radio" && status) {
      this.setState({
        status: false
      });
    }

    if (type === "checkbox") {
      this.setState({
        value: { ...this.state.value, [value]: checked }
      });
    }

    if (type === "fill" || type === "radio") {
      this.setState({ value });
    }
  };

  handleCheckAnswer = () => {
    const { correctAnswer, type } = this.props;
    const { value } = this.state;

    if (type === "fill" || type === "radio") {
      const status = value.toLowerCase() === correctAnswer.toLowerCase();
      this.setState({
        status
      });
    }

    if (type === "checkbox") {
      const currentAnswer = Object.keys(value).filter(answer => value[answer]);

      const isCorrect = currentAnswer.every(a => correctAnswer.indexOf(a) > -1);

      this.setState({
        status: isCorrect
      });
    }

    this.setState({
      disabled: true
    });
  };

  render() {
    const { text, description, answers, type, correctAnswer } = this.props;
    const { value, status, disabled } = this.state;

    return (
      <StyledCard>
        <Card.Content>
          <Card.Header>{text}</Card.Header>
          <Card.Description>
            {type !== "fill"
              ? "Check that you consider correct."
              : "Fill in this field."}
          </Card.Description>
        </Card.Content>
        <Card.Content>
          answers, value, status, correctAnswer, handleChange
          <StyledForm>
            <QuestionType
              type={type}
              value={value}
              status={status}
              correctAnswer={correctAnswer}
              handleChange={this.handleChange}
            />
          </StyledForm>
        </Card.Content>
        <ButtonContent>
          <Button
            secondary
            onClick={this.handleCheckAnswer}
            disabled={disabled}
          >
            Ok
          </Button>
        </ButtonContent>
      </StyledCard>
    );
  }
}

Question.defaultProps = {
  type: "fill"
};

export default Question;
