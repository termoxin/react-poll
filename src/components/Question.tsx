import React, { Component } from "react";
import { Form, Card, Icon, Radio, Button } from "semantic-ui-react";
import styled from "styled-components";
import QuestionType from "./QuestionType";
import merge from "../helpers/merge";

const ButtonContent = styled(Card.Content)`
  display: flex;
  justify-content: flex-end;
`;

const StyledCard = styled(Card)`
  && {
    width: 600px;
    min-width: 320px;
  }
`;

const CardHeader = styled(Card.Header)`
  && {
    display: flex;
    justify-content: space-evenly;
  }
`;

const StyledForm = styled(Form)`
  display: flex;
  flex-flow: column nowrap;
  align-items: flex-start;
`;

const CardTextID = styled.span`
  font-size: 13px;
  font-weight: 400;
`;

const CardTextP = styled.p``;

Button.displayName = "Button";

interface Props {
  answers: Array<object>;
  correctAnswer: string;
  handleChecking: Function;
  id: number | string;
  text: string;
  type: string;
  description: string;
  handleChange: Function;
  count: number;
}

interface State {
  inputValue: string | object;
  status: boolean;
  disabled?: boolean;
}

interface CheckboxProps {
  value: string;
  checked: boolean;
}

class Question extends Component<Props, State> {
  static defaultProps: { type: string; disabled: boolean } = {
    type: "fill",
    disabled: false
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      inputValue: "",
      status: null,
      disabled: false
    };
  }

  handleChange = (
    e: React.ChangeEvent<HTMLInputElement>,
    data: CheckboxProps
  ) => {
    const { type } = this.props;
    const { status, inputValue } = this.state;

    if (data) {
      const { value, checked } = data;

      if (type === "radio" && status) {
        this.setState({
          status: false
        });
      }

      if (type === "checkbox") {
        this.setState({
          inputValue: merge(inputValue, { [value]: checked })
        });
      }

      if (type === "fill" || type === "radio") {
        this.setState({ inputValue: value });
      }
    }
  };

  handleCheckAnswer = () => {
    const { correctAnswer, type, handleChecking, text } = this.props;
    const { inputValue, status } = this.state;
    const handleCheckingOrPureFunction = handleChecking || function() {};

    if (type === "fill" || type === "radio") {
      const iv = inputValue.toString().toLowerCase();
      let ca =
        typeof correctAnswer !== "undefined" ? correctAnswer.toLowerCase() : "";

      const status = iv === ca;
      this.setState({
        status
      });
    }

    if (type === "checkbox") {
      const currentAnswer = Object.keys(inputValue).filter(
        answer => inputValue[answer]
      );

      if (currentAnswer.length !== correctAnswer.length) {
        this.setState({ status: false });
      } else {
        const isCorrect = currentAnswer.every(
          a => correctAnswer.indexOf(a) > -1
        );

        this.setState({
          status: isCorrect
        });
      }
    }

    this.setState(
      {
        disabled: true
      },
      () => handleCheckingOrPureFunction(this.state.status, text)
    );
  };

  render() {
    const { text, answers, type, correctAnswer, id, count } = this.props;
    const { inputValue, status, disabled } = this.state;

    return (
      <StyledCard>
        <Card.Content>
          <Card.Header>
            <CardTextID>
              {id}/{count}
            </CardTextID>
            <CardTextP>{text}</CardTextP>
          </Card.Header>
          <Card.Description>
            {type !== "fill"
              ? "Check that you consider correct."
              : "Fill in this field."}
          </Card.Description>
        </Card.Content>
        <Card.Content>
          <StyledForm>
            <QuestionType
              type={type}
              value={inputValue}
              answers={answers}
              status={status}
              disabled={disabled}
              correctAnswer={correctAnswer}
              handleChange={this.handleChange}
            />
          </StyledForm>
        </Card.Content>
        <ButtonContent>
          <Button
            secondary
            onClick={this.handleCheckAnswer}
            disabled={disabled}
          >
            Ok
          </Button>
        </ButtonContent>
      </StyledCard>
    );
  }
}

export default Question;
