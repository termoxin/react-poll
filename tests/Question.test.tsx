import React from "react";
import { create } from "react-test-renderer";
import Question from "../src/components/Question";

const answers = [
  { id: 1, text: "Yeah!" },
  { id: 2, text: "No!" },
  { id: 3, text: "Probably." }
];

describe("Question component", () => {
  // @TODO: If you are not make sure that it won't change not it'll launch

  it("matches the snapshot", () => {
    const tree = create(<Question />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("disabled should be true when button is clicked", () => {
    const component = create(<Question />);
    const rootInstance = component.root;
    const button = rootInstance.findByType("button");

    button.props.onClick();

    expect(button.props.disabled).toEqual(true);
  });

  it("question index is 5", () => {
    const component = create(<Question id={5} count={10} />);
    const rootInstance = component.root;
    const span = rootInstance.findByType("span");
    const index = span.props.children.join("");

    expect(index).toEqual("5/10");
  });

  it("question index is 'Text'", () => {
    const component = create(<Question text="Text" />);
    const rootInstance = component.root;
    const span = rootInstance.findByType("p");
    const text = span.props.children;

    expect(text).toEqual("Text");
  });

  it("there is right input type and type is equal checkbox", () => {
    const component = create(
      <Question type="checkbox" answers={answers} correctAnswer={["Yeah!"]} />
    );

    const root = component.root;
    const listOfCheckboxes = root.findAll(element => element.type === "input");

    expect(listOfCheckboxes[0].props.value).toBe("Yeah!");
    expect(listOfCheckboxes[2].props.value).toBe("Probably.");
  });

  it("there is right input type and type is equal fill", () => {
    const component = create(<Question correctAnswer="TEXT" />);
    const root = component.root;
    const input = root.findByType("input").props.type;

    expect(input).toEqual("text");
  });

  it("there is right input type and type is equal radio", () => {
    const component = create(
      <Question type="radio" answers={answers} correctAnswer="TEXT" />
    );
    const root = component.root;
    const listOfRadios = root.findAll(element => element.type === "input");

    expect(listOfRadios[0].props.value).toBe("Yeah!");
    expect(listOfRadios[2].props.value).toBe("Probably.");
  });

  it("successful checking of correct answer in input type such as fill", () => {
    const component = create(<Question correctAnswer="TEXT" />);
    const root = component.root;
    const instance = component.getInstance();
    const input = root.findByType("input");
    const button = root.findByType("button");

    instance.setState({ inputValue: "TEXT" });

    button.props.onClick();

    expect(input.props.disabled).toEqual(true);
    expect(instance.state.inputValue).toEqual("TEXT");
    expect(instance.state.status).toEqual(true);
  });

  it("successful checking of correct answer in input type such as radio", () => {
    const component = create(
      <Question correctAnswer="TEXT" type="radio" answers={answers} />
    );
    const root = component.root;
    const instance = component.getInstance();
    const listOfRadios = root.findAll(element => element.type === "input");
    const button = root.findByType("button");

    instance.setState({ inputValue: "TEXT" });

    button.props.onClick();

    expect(instance.state.inputValue).toEqual("TEXT");
    expect(instance.state.status).toEqual(true);
  });

  it("changing state and value when onChange in fill", () => {
    const component = create(<Question correctAnswer="TEXT" />);
    const root = component.root;
    const instance = component.getInstance();
    const event = { target: { value: "TEXT" } };
    const input = root.findByType("input");

    input.props.onChange(event);

    expect(instance.state.inputValue).toEqual("TEXT");
  });

  it("changing state and value when onChange in radio", () => {
    const component = create(
      <Question correctAnswer="Yeah!" type="radio" answers={answers} />
    );
    const root = component.root;
    const instance = component.getInstance();
    const event = { value: "TEXT" };
    const input = root.findAll(element => element.props.onChange);

    input[1].props.onChange({}, event);

    expect(instance.state.inputValue).toEqual("TEXT");
  });

  it("changing state and value when onChange in checkbox", () => {
    const component = create(
      <Question correctAnswer={["Yeah"]} type="checkbox" answers={answers} />
    );
    const root = component.root;
    const instance = component.getInstance();
    const fEvent = { value: "Yeah", checked: true };
    const sEvent = { value: "Probably.", checked: true };
    const input = root.findAll(element => element.props.onChange);

    input[0].props.onChange({}, fEvent);
    input[2].props.onChange({}, sEvent);

    expect(instance.state.inputValue).toEqual({
      "Probably.": true,
      Yeah: true
    });
  });
});
