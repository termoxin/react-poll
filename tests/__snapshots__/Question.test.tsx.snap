// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question component matches the snapshot 1`] = `
<div
  className="ui card StyledCard-hvznlg-1 fKVLbp"
  onClick={[Function]}
>
  <div
    className="content"
  >
    <div
      className="header"
    >
      <span
        className="CardTextID-hvznlg-4 ixDgNO"
      >
        /
      </span>
      <p
        className="CardTextP-hvznlg-5 faoLx"
      />
    </div>
    <div
      className="description"
    >
      Fill in this field.
    </div>
  </div>
  <div
    className="content"
  >
    <form
      className="ui form StyledForm-hvznlg-3 bUFIME"
      onSubmit={[Function]}
    >
      <div
        className="ui input StyledInput-sc-16u23nf-0 gXWRui"
        status="false"
      >
        <input
          disabled={false}
          onChange={[Function]}
          placeholder="Type your answer here"
          type="text"
        />
      </div>
    </form>
  </div>
  <div
    className="content ButtonContent-hvznlg-0 bYUtbp"
  >
    <button
      className="ui secondary button"
      onClick={[Function]}
    >
      Ok
    </button>
  </div>
</div>
`;
